README.md
# Prometheus exporter for Druid (http://druid.io/)

Collects HTTP POST JSON data coming from Druid daemons and expose them formatted
following the [Prometheus](https://prometheus.io) standards.

## Running

The easiest way to run `druid_exporter` is via a virtualenv:

```
  virtualenv .venv
  .venv/bin/python setup.py install
  .venv/bin/druid_exporter path-to-metrics-config-file.json
```

By default metrics are exposed on TCP port `8000`. Python 2 is not supported.

### Running with Docker

Issue the following command to build the image. Default base image (python:3.8-alpine) can be replaced.
```bash
docker build -t druid-exporter:latest .
docker build -t druid-exporter:latest --build-arg BASE_IMAGE=<python-image> .
```

Issue the following command to run.
```bash
docker run -it -d -p 8000:8000 druid-exporter:latest conf/example_druid_0_12_3.json
```

Example `docker-compose.yml`:
```yaml
druid-exporter:
  image: "druid-exporter:latest"
  command: 
    - "conf/example_druid_0_12_3.json"
  ports:
    - published: 8000
      target: 8000
```

## Druid versions supported

This exporter is tested and used by the Wikimedia foundation with Druid version 0.12.3,
so it might not work as expected for newer versions. The exporter is designed in such a way
that metrics are configured entirely by admin/users, so it should be resilient to future
versions of Druid.

## How does it work?

Druid can be configured to emit metrics (JSON data) to a HTTP endpoint configured
via the following runtime properties:

http://druid.io/docs/0.12.3/configuration/index.html#http-emitter-module

The druid prometheus exporter accepts HTTP POST data, inspects it and stores/aggregates
every supported datapoint into a data structure. It then formats the
data on the fly to Prometheus metrics when a GET /metrics is requested.

The exporter is also able to pull metrics from a Kafka topic, when Druid is configured to do so:

https://druid.apache.org/docs/latest/development/extensions-contrib/kafka-emitter.html

The Kafka support is optional and should be used when the number of metrics/datapoints emitted
by Druid is very high (rule of thumb could be around 1000 datapoints/s). Please check the section
about performance considerations below for more info.

This exporter is supposed to be run on each host running one or multiple Druid daemons.

## Supported metrics and labels

In theory any metric that Druid emits should be supported by this exporter, even if there might
be some use cases not covered (if it happens to you, please report the problem!).
As example, let's start with the following metric emitted by druid:

```
[{"feed":"metrics","timestamp":"2020-04-23T11:36:27.866Z","service":"druid/peon","host":"druid1001.eqiad.wmnet:8201",
"version":"0.12.3","metric":"ingest/events/unparseable","value":0,"dataSource":"wmf_netflow",
"taskId":["index_kafka_wmf_netflow_ee74c5a5820837c_hedlopdm"]}]
```

The exporter doesn't know anything about what metrics it should support, everything is stated
in the metrics JSON config file that must be provided as parameter. In this case, we could add the following
entry to the JSON to support the metric:

```
[..]
    "peon": {
    	[..]
        "ingest/events/unparseable": {
            "prometheus_metric_name": "druid_realtime_ingest_events_unparseable_count",
            "type": "gauge",
            "labels": ["dataSource"],
            "description": "Number of events rejected because the events are unparseable."
        },
        [..]
[..]
```

Pleae note that the name of the label corresponds to the field in the "feed" event emitted
by Druid.
If you want to check a more complete example, please see the conf directory of this repository.

Please check the following document for more info about metrics emitted by Druid:
http://druid.io/docs/0.12.3/operations/metrics.html

The JVM metrics are currently not supported, please check other projects
like https://github.com/prometheus/jmx_exporter if you need to collect them.

## Known limitations

When a Druid cluster is running with multiple coordinators or overlords,
only one of them acts as leader and the others are in standby mode. From the metrics
emission point of view, this means that only the daemon acting as leader emits druid metrics,
the others don't (except for metrics like Jetty ones). This has caused some confusion
in our day to day operations, since when a coordinator or a overlord leader looses
its status in favor of another one (for example due to a restart) it stops emitting
metrics, and the Prometheus exporter keeps reporting the last known state. This might
be confusing to see at first (expecially if metrics are aggregated) so the current
"fix" is to restart the Druid Prometheus exporter when a coordinator or a overlord
leader are restarted. Future versions of this project might contain a fix, pull
requests are welcome!

## Performance considerations

In https://github.com/wikimedia/operations-software-druid_exporter/issues/11 some users
brought up an interesting use case for the exporter, namely handling peaks of 1500 datapoints/s
sent from Druid brokers (one can check the rate of datapoints/s via the
`druid_exporter_datapoints_registered_total metric`).

The exporter's code was refactored to be able to scale more
(if you are curious, check [this commit](https://github.com/wikimedia/operations-software-druid_exporter/commit/f22c6d9f8707ae2d274db9b10669b971beed64ab)), but it wasn't enough, since users kept reporting timeouts from Druid daemons while sending
datapoints to the exporter. There are some recommendations to follow:
* Try to use a dedicated exporter instance for daemons sending high volumes of datapoints.
* Try to tune the HTTP emitter's settings via https://druid.apache.org/docs/latest/configuration/index.html#http-emitter-module
  (if the Druid version that you are running supports them).
If the above is not enough, the exporter can be configured to also pull datapoints from
a Kafka topic (see [https://druid.apache.org/docs/latest/development/extensions-contrib/kafka-emitter.html](Kafka)). With this configuration, the exporter will ingest datapoints coming via
HTTP and Kafka at the same time. An ideal solution is to force Druid daemons emitting too many
datapoints/s to use the KafkaEmitter, and the other ones to use the HTTPEmitter.

debian/changelog
prometheus-druid-exporter (0.8-2) bullseye-wikimedia; urgency=medium

  * Rebuild for bullseye

 -- Moritz Mühlenhoff <moritz@wikimedia.org>  Thu, 23 Mar 2023 09:35:21 +0100

prometheus-druid-exporter (0.8-1) unstable; urgency=low

  * Upgrade to upstream version 0.8.
  * Require a more recent version of python3-python-client.

 -- Luca Toscano <ltoscano@wikimedia.org>  Sat, 02 May 2020 10:53:36 +0000

prometheus-druid-exporter (0.6-1) unstable; urgency=low

  * Allow a metric to be emitted by multiple daemons,
    even with different metrics. The use case of 'segment/count'
    is unique since the metric is emitted by both coordinator and
    historical, with different labels.
  * Correct the historical's metric labels for segment/count
    and segment/used.
  * Change name of druid_(broker|historical)_query_cache_sizebytes_count
    to a more appropriate _query_cache_size_bytes
  * No-op variable name changes to be more consistent in the code.
  * Prefer absolute import for the collector.py module
  * Add a more precise and robust set of tests.

 -- Luca Toscano <ltoscano@wikimedia.org>  Fri, 08 Dec 2017 10:53:36 +0000

prometheus-druid-exporter (0.5-1) unstable; urgency=low

  * Add support for realtime ingestion metrics
  * Removed unused --allow-daemons parameter.

 -- Luca Toscano <ltoscano@wikimedia.org>  Wed, 06 Dec 2017 10:53:36 +0000

prometheus-druid-exporter (0.4-1) unstable; urgency=low

  * Remove unused query/node/* metrics support.

 -- Luca Toscano <ltoscano@wikimedia.org>  Tue, 21 Nov 2017 10:53:36 +0000

prometheus-druid-exporter (0.3-1) unstable; urgency=low

  * Fix decoding of JSON data in python3
  * Drop support for Python 2 to simplify the code maintenance.
  * Add the --encoding parameter (default to utf-8) to allow a user
    to instruct what encoding the druid exporter should expect to use
    when decoding POST data coming from Druid. The documentation
    seems to suggest that UTF-8 is the preferred one.

 -- Luca Toscano <ltoscano@wikimedia.org>  Mon, 20 Nov 2017 12:53:36 +0000

prometheus-druid-exporter (0.2-1) unstable; urgency=low

  * Fix import of local libraries for python3

 -- Luca Toscano <ltoscano@wikimedia.org>  Mon, 20 Nov 2017 10:53:36 +0000

prometheus-druid-exporter (0.1-1) unstable; urgency=low

  * Initial packaging

 -- Luca Toscano <ltoscano@wikimedia.org>  Fri, 06 Oct 2017 13:53:36 +0000


druid_exporter/exporter.py
#!/usr/bin/python
# Copyright 2017 Luca Toscano
#                Filippo Giunchedi
#                Wikimedia Foundation
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse
import json
import logging
import sys

from druid_exporter import collector
from prometheus_client import generate_latest, make_wsgi_app, REGISTRY
from gevent.pywsgi import WSGIServer
from wsgigzip import gzip

log = logging.getLogger(__name__)

class DruidWSGIApp(object):
    def __init__(self, post_uri, druid_collector, prometheus_app, encoding):
        self.prometheus_app = prometheus_app
        self.druid_collector = druid_collector
        self.post_uri = post_uri
        self.encoding = encoding

    def __call__(self, environ, start_response):
        if (environ['REQUEST_METHOD'] == 'GET' and
                environ['PATH_INFO'] == '/metrics'):
            return self.prometheus_app(environ, start_response)
        elif (environ['REQUEST_METHOD'] == 'POST' and
                environ['PATH_INFO'] == self.post_uri and
                environ['CONTENT_TYPE'] == 'application/json'):
            try:
                request_body_size = int(environ.get('CONTENT_LENGTH', 0))
                request_body = environ['wsgi.input'].read(request_body_size)
                datapoints = json.loads(request_body.decode(self.encoding))
                # The HTTP metrics emitter can batch datapoints and send them to
                # a specific endpoint stated in the logs (this tool).
                for datapoint in datapoints:
                    log.debug("Processing datapoint: {}".format(datapoint))
                    self.druid_collector.datapoints_queue.put(datapoint)
                status = '200 OK'
            except Exception as e:
                log.exception('Error while processing the following POST data')
                status = '400 Bad Request'
        else:
            status = '400 Bad Request'
        start_response(status, [])
        return ''


def with_gzip_compression(handle_cb):
    @gzip(compress_level=6)
    def handle(environ, start_response):
        return handle_cb(environ, start_response)

    return handle


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('-l', '--listen', metavar='ADDRESS',
                        help='Listen on this address', default=':8000')
    parser.add_argument('-u', '--uri', default='/',
                        help='The URI to check for POSTs coming from Druid')
    parser.add_argument('-d', '--debug', action='store_true',
                        help='Enable debug logging')
    parser.add_argument('-e', '--encoding', default='utf-8',
                        help='Encoding of the Druid POST JSON data.')
    args = parser.parse_args()

    if args.debug:
        logging.basicConfig(level=logging.DEBUG)
    else:
        logging.basicConfig(level=logging.WARNING)

    collect_metrics_from = []

    address, port = args.listen.split(':', 1)
    log.info('Starting druid_exporter on %s:%s', address, port)

    druid_collector = collector.DruidCollector()
    prometheus_app = make_wsgi_app()
    druid_wsgi_app = DruidWSGIApp(args.uri, druid_collector,
                                  prometheus_app, args.encoding)

    httpd = WSGIServer((address, int(port)), with_gzip_compression(druid_wsgi_app))
    httpd.serve_forever()

if __name__ == "__main__":
    sys.exit(main())


druid_exporter/collector.py
# Copyright 2017 Luca Toscano
#                Filippo Giunchedi
#                Wikimedia Foundation
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import re
import logging
import queue
import threading

from prometheus_client.core import (Counter, Gauge, Histogram, Summary)

log = logging.getLogger(__name__)

SKIP_METRIC = {
    'type': 'skip',
}

QUERY_TIME_SUMMARY_METRIC = {
    'labels': ['type', 'dataSource'],
    'suffix': '_ms',
    'type': 'summary',
}

QUERY_NODE_TIME_METRIC = {
    'labels': ['type', 'dataSource', 'server'],
    'suffix': '_ms',
    'type': 'summary',
}

SEGMENT_TIER_METRIC = {
    'labels': ['tier'],
}

SEGMENT_SOURCE_METRIC = {
    'labels': ['dataSource'],
}

SEGMENT_SERVER_METRIC = {
    'labels': ['server'],
}

SEGMENT_USED_METRIC = {
    'labels': ['dataSource', 'tier', 'priority'],
}

INGEST_METRIC = {
    'labels': ['dataSource'],
    'type': 'counter',
}

INGEST_TIME_METRIC = {
    'labels': ['dataSource'],
    'suffix': '_ms',
    'type': 'summary',
}

JETTY_METRICS = {
    'jetty/numOpenConnections': {},
}

QUERY_METRICS = {
    'query/bytes': {'labels': ['type', 'dataSource'], 'type': 'summary'},
    'query/cpu/time': {**QUERY_TIME_SUMMARY_METRIC},
    'query/time': {'labels': ['type', 'dataSource'], 'suffix': '_ms', 'type': 'histogram', 'buckets': (150, 1000, 5000, 10000, 30000, float('inf'))},
}

QUERY_CACHE_METRICS = {
    'query/cache/caffeine/delta/evictionBytes': {**SKIP_METRIC},
    'query/cache/caffeine/delta/loadTime': {**SKIP_METRIC},
    'query/cache/caffeine/delta/requests': {**SKIP_METRIC},
    'query/cache/caffeine/total/evictionBytes': {},
    'query/cache/caffeine/total/loadTime': {},
    'query/cache/caffeine/total/requests': {},
    'query/cache/delta/averageBytes': {**SKIP_METRIC},
    'query/cache/delta/errors': {**SKIP_METRIC},
    'query/cache/delta/evictions': {**SKIP_METRIC},
    'query/cache/delta/hitRate': {**SKIP_METRIC},
    'query/cache/delta/hits': {**SKIP_METRIC},
    'query/cache/delta/misses': {**SKIP_METRIC},
    'query/cache/delta/numEntries': {**SKIP_METRIC},
    'query/cache/delta/sizeBytes': {**SKIP_METRIC},
    'query/cache/delta/timeouts': {**SKIP_METRIC},
    'query/cache/total/averageBytes': {},
    'query/cache/total/errors': {},
    'query/cache/total/evictions': {},
    'query/cache/total/hitRate': {},
    'query/cache/total/hits': {},
    'query/cache/total/misses': {},
    'query/cache/total/numEntries': {},
    'query/cache/total/sizeBytes': {},
    'query/cache/total/timeouts': {},
}

QUERY_COUNT_METRIC = {
    'type': 'counter',
    'suffix': '',
}

QUERY_COUNT_STATS = {
    'query/failed/count': {**QUERY_COUNT_METRIC},
    'query/interrupted/count': {**QUERY_COUNT_METRIC},
    'query/success/count': {**QUERY_COUNT_METRIC},
}

SEGMENT_METRICS = {
    'query/segment/time': {**QUERY_TIME_SUMMARY_METRIC},
    'query/segmentAndCache/time': {**QUERY_TIME_SUMMARY_METRIC},
    'query/wait/time': {**QUERY_TIME_SUMMARY_METRIC},
    'segment/max': {'suffix': '_bytes'},
    'segment/scan/pending': {},
    'segment/used': {**SEGMENT_USED_METRIC, 'suffix': '_bytes'},
    'segment/usedPercent': {**SEGMENT_USED_METRIC, 'suffix': '_percent'},
    'segment/count': {**SEGMENT_USED_METRIC},
}

SEGMENT_COORDINATOR_METRICS = {
    'segment/assigned/count': {**SEGMENT_TIER_METRIC},
    'segment/deleted/count': {**SEGMENT_TIER_METRIC},
    'segment/dropped/count': {**SEGMENT_TIER_METRIC},
    'segment/moved/count': {**SEGMENT_TIER_METRIC},
    'segment/unneeded/count': {**SEGMENT_TIER_METRIC},

    'segment/cost/normalization': {**SEGMENT_TIER_METRIC, 'suffix': ''},
    'segment/cost/normalized': {**SEGMENT_TIER_METRIC, 'suffix': ''},
    'segment/cost/raw': {**SEGMENT_TIER_METRIC, 'suffix': ''},

    'segment/loadQueue/size': {**SEGMENT_SERVER_METRIC, 'suffix': '_bytes'},
    'segment/loadQueue/failed': {**SEGMENT_SERVER_METRIC},
    'segment/loadQueue/count': {**SEGMENT_SERVER_METRIC},
    'segment/dropQueue/count': {**SEGMENT_SERVER_METRIC},

    'segment/size': {**SEGMENT_SOURCE_METRIC, 'suffix': '_bytes'},
    'segment/count': {**SEGMENT_SOURCE_METRIC},
    'segment/unavailable/count': {**SEGMENT_SOURCE_METRIC},

    'segment/overShadowed/count': {},
    'segment/underReplicated/count': {'labels': ['tier', 'dataSource']},
}

BROKER_METRICS = {
    'avatica/remote/JsonHandler/Handler/Serialization': {**SKIP_METRIC},
    'avatica/server/AvaticaJsonHandler/Handler/RequestTimings': {**SKIP_METRIC},
    'query/intervalChunk/time': {**QUERY_TIME_SUMMARY_METRIC},
    'query/node/bytes': {'labels': ['type', 'dataSource', 'server'], 'type': 'summary'},
    'query/node/time': {**QUERY_NODE_TIME_METRIC},
    'query/node/ttfb': {**QUERY_NODE_TIME_METRIC},
}

INGEST_METRICS = {
    'ingest/events/messageGap': {'labels': ['dataSource'], 'suffix': '_ms'},
    'ingest/events/processed': {**INGEST_METRIC},
    'ingest/events/thrownAway': {**INGEST_METRIC},
    'ingest/events/unparseable': {**INGEST_METRIC},
    'ingest/handoff/count': {**INGEST_METRIC},
    'ingest/handoff/failed': {**INGEST_METRIC},
    'ingest/kafka/lag': {'labels': ['dataSource'], 'suffix': '_size'},
    'ingest/merge/cpu': {**SKIP_METRIC},
    'ingest/merge/time': {**INGEST_TIME_METRIC},
    'ingest/persists/backPressure': {**INGEST_TIME_METRIC},
    'ingest/persists/count': {**INGEST_METRIC},
    'ingest/persists/cpu': {**SKIP_METRIC},
    'ingest/persists/failed': {**INGEST_METRIC},
    'ingest/persists/time': {**INGEST_TIME_METRIC},
    'ingest/rows/output': {**INGEST_METRIC},
    'ingest/sink/count': {**INGEST_METRIC},
}

JVM_POOL_METRIC = {
    'labels': ['poolKind', 'poolName'],
    'type': 'gauge',
    'suffix': '_bytes',
}

# these metrics are reported by the JvmMonitor option
JVMMONITOR_METRICS = {
    'jvm/pool/committed':   {**JVM_POOL_METRIC},
    'jvm/pool/init':        {**JVM_POOL_METRIC},
    'jvm/pool/max':         {**JVM_POOL_METRIC},
    'jvm/pool/used':        {**JVM_POOL_METRIC},

    'jvm/bufferpool/count':       { 'labels': [ 'bufferpoolName' ], 'suffix': '' },
    'jvm/bufferpool/used':        { 'labels': [ 'bufferpoolName' ], 'suffix': '_bytes' },
    'jvm/bufferpool/capacity':    { 'labels': [ 'bufferpoolName' ], 'suffix': '' },

    'jvm/mem/committed':  { 'labels': [ 'memKind' ], 'suffix': '_bytes' },
    'jvm/mem/init':       { 'labels': [ 'memKind' ], 'suffix': '_bytes' },
    'jvm/mem/max':        { 'labels': [ 'memKind' ], 'suffix': '_bytes' },
    'jvm/mem/used':       { 'labels': [ 'memKind' ], 'suffix': '_bytes' },

    'jvm/gc/count': { 'labels': [ 'gcName', 'gcGen' ], 'suffix': '' },
    'jvm/gc/cpu':   { 'labels': [ 'gcName', 'gcGen' ], 'suffix': '_ns' },
}


MINUTE = 60000

INDEXING_SERVICE_METRICS = {
    'task/run/time': {'labels': ['dataSource', 'taskType', 'taskStatus'], 'suffix': '_ms', 'type': 'histogram', 'buckets': (30 * MINUTE, 60 * MINUTE, 90 * MINUTE, 120 * MINUTE, float('inf'))},
    'segment/added/bytes': {'labels': ['dataSource', 'taskType'], 'type': 'counter'},
}

class DruidCollector(object):
    datapoints_processed = Counter('druid_exporter_datapoints_processed_count', '')

    def __init__(self):
        self.datapoints_queue = queue.Queue()
        threading.Thread(target=self.process_queued_datapoints).start()
        
        self.supported_metrics = {
            'coordinator': {
                **JETTY_METRICS,
                **SEGMENT_COORDINATOR_METRICS,
            },
            'broker': {
                **JETTY_METRICS,
                **QUERY_METRICS,
                **QUERY_CACHE_METRICS,
                **QUERY_COUNT_STATS,
                **BROKER_METRICS,
                **SEGMENT_METRICS,
            },
            'historical': {
                **JETTY_METRICS,
                **QUERY_METRICS,
                **QUERY_CACHE_METRICS,
                **QUERY_COUNT_STATS,
                **SEGMENT_METRICS,
            },
            'overlord': {
                **JETTY_METRICS,
                **INGEST_METRICS,
                **INDEXING_SERVICE_METRICS,
            },
            'middlemanager': {
                **JETTY_METRICS,
            },
            'peon': {
                **JETTY_METRICS,
                **QUERY_METRICS,
                **QUERY_CACHE_METRICS,
                **QUERY_COUNT_STATS,
                **INGEST_METRICS,
                **SEGMENT_METRICS,
                **JVMMONITOR_METRICS,
            },
            'realtime': {
                **JETTY_METRICS,
                **QUERY_METRICS,
                **QUERY_CACHE_METRICS,
                **QUERY_COUNT_STATS,
                **INGEST_METRICS,
            },
        }

    def _get_metric_name(self, daemon, metric_name, config):
        if 'name' in config:
            metric_name = config['name']

        log.debug("initial metric_name: {}".format(metric_name))

        metric_name = re.sub('/', r'_', metric_name)
        metric_name = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', metric_name)
        metric_name = re.sub('([a-z0-9])([A-Z])', r'\1_\2', metric_name).lower()

        if len(config['suffix']) > 0 and metric_name.endswith(config['suffix']):
            pass
        elif metric_name.endswith('_bytes'):
            pass
        elif metric_name.endswith('_count'):
            pass
        else:
            metric_name = metric_name + config['suffix']

        log.debug("final metric_name: {}".format(metric_name))

        return 'druid_' + daemon + '_' + metric_name
    
    def process_queued_datapoints(self):
        while True:
            datapoint = self.datapoints_queue.get()
            self.process_datapoint(datapoint)

    def process_datapoint(self, datapoint):
        if (datapoint['feed'] != 'metrics'):
            log.debug("'feed' field is not 'metrics' in datapoint, skipping: {}".format(datapoint))
            return

        daemon = str(datapoint['service']).replace('druid/', '').lower()

        if (daemon not in self.supported_metrics):
            log.debug("daemon '{}' is not supported, skipping: {}".format(daemon, datapoint))
            return

        metric_name = str(datapoint['metric'])

        if (metric_name not in self.supported_metrics[daemon]):
            log.debug("metric '{}' is not supported, skipping: {}".format(datapoint['metric'], datapoint))
            return

        config = self.supported_metrics[daemon][metric_name]
        config.setdefault('labels', [])
        config.setdefault('type', 'gauge')
        config.setdefault('suffix', '_count')

        metric_type = config['type']

        if metric_type == 'skip':
            return

        metric_name = self._get_metric_name(daemon, metric_name, config)
        metric_value = float(datapoint['value'])
        metric_labels = tuple(sorted(config['labels'] + ['host']))
        label_values = tuple([datapoint[label_name] for label_name in metric_labels])

        if '_metric_' not in config:
            if metric_type == 'counter':
                config['_metric_'] = Counter(metric_name, metric_name, metric_labels)
            if metric_type == 'gauge':
                config['_metric_'] = Gauge(metric_name, metric_name, metric_labels)
            elif metric_type == 'summary':
                config['_metric_'] = Summary(metric_name, metric_name, metric_labels)
            elif metric_type == 'histogram':
                config['_metric_'] = Histogram(metric_name, metric_name, metric_labels, buckets=config['buckets'])

        metric = config['_metric_']

        if len(metric_labels) > 0:
            metric = metric.labels(*label_values)

        if metric_type == 'counter':
            metric.inc(metric_value)
        if metric_type == 'gauge':
            metric.set(metric_value)
        elif metric_type == 'summary':
            metric.observe(metric_value)
        elif metric_type == 'histogram':
            metric.observe(metric_value)

        self.datapoints_processed.inc()


utils/parse_failed_log.py
#!/usr/bin/python
# Copyright 2020 Rafa Diaz
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# All the failed metris are written in the log with a similar format:
#
# DEBUG:druid_exporter.collector:The following datapoint is not supported, either because the 'feed' field is not 'metrics' or the metric itself is not supported: {'feed': 'metrics', 'timestamp': '2020-05-07T09:57:08.983Z', 'service': 'druid/router', 'host': '10.132.0.27:8888', 'version': '0.18.0', 'metric': 'query/time', 'value': 7, 'context': {'queryId': '2c08a196-9c11-49c3-a5e7-b41357388f24', 'timeout': 40000}, 'dataSource': 'daily_page_export', 'duration': 'PT9223372036854775.807S', 'hasFilters': 'false', 'id': '2c08a196-9c11-49c3-a5e7-b41357388f24', 'interval': ['-146136543-09-08T08:23:32.096Z/146140482-04-24T15:36:27.903Z'], 'remoteAddress': '10.0.3.217', 'success': 'true', 'type': 'timeBoundary'}
# 
# [{"feed":"metrics","timestamp":"2020-04-23T11:36:27.866Z","service":"druid/peon","host":"druid1001.eqiad.wmnet:8201",
# "version":"0.12.3","metric":"ingest/events/unparseable","value":0,"dataSource":"wmf_netflow",
# "taskId":["index_kafka_wmf_netflow_ee74c5a5820837c_hedlopdm"]}]
#
# Which should be convert in something like this:
# [..]
#     "peon": {
#     	[..]
#         "ingest/events/unparseable": {
#             "prometheus_metric_name": "druid_realtime_ingest_events_unparseable_count",
#             "type": "gauge",
#             "labels": ["dataSource"],
#             "description": "Number of events rejected because the events are unparseable."
#         },
#         [..]
# [..]


import argparse
import json
import logging
import sys
import re

log = logging.getLogger(__name__)


def load_descriptions():
    desc_dict = {}

    fd = open('descriptions.txt', 'r') 
    lines = fd.readlines()
    
    # Strips the newline character 
    for line in lines:
        desc_match = re.match("(.*)\t(.*)",line) 
        if (desc_match):
            # print("Match")
            desc_dict[desc_match.group(1)] = desc_match.group(2)
    return desc_dict

def is_processed(service, metric):
    if service not in processed_metrics:
        processed_metrics[service] = []
        processed_metrics[service].append(metric)
        return False
    if metric not in processed_metrics[service]:
        processed_metrics[service].append(metric)
        return False

    return True


p = re.compile("DEBUG:druid_exporter.collector:The following datapoint is not supported, either because the 'feed' field is not 'metrics' or the metric itself is not supported: (.*)")

# Descriptions for metrics are taken from: https://druid.apache.org/docs/latest/operations/metrics.html
descriptions = load_descriptions()
# print(descriptions)

processed_metrics = {}

for line in sys.stdin:
    unsupported_metric = p.search(line)

    if unsupported_metric:
        metric = unsupported_metric.group(1)
        parsed_json = json.loads(metric.replace("'", '"'))
        # print("found: {0}".format(unsupported_metric.group(1)))
        
        if is_processed(parsed_json['service'], parsed_json['metric']):
            continue

        print("\n\n----------------")
        print("Found metric: ")
        print(json.dumps(parsed_json, indent=4, sort_keys=True))
    else:
        continue

    # print("matches: {1}".format(result.group(0)))
    print("\nThis is a metric for: {0}".format(parsed_json['service']))
    output_json = {}
    metric_name = parsed_json['metric']
    output_json[metric_name] = {}
    output_json[metric_name]["prometheus_metric_name"] = "druid_{0}".format(metric_name.replace('/','_'))
    # So far we don't have better way to guess a right description so this can be a temporary solution
    if (metric_name in descriptions.keys()):
        output_json[metric_name]["description"] = descriptions[metric_name]
    else:
        output_json[metric_name]["description"] = "druid_{0}".format(metric_name.replace('/','_'))
    

    # Some heuristic to guess the labels
    possible_labels = ["dataSource", "memKind", "bufferpoolName"]
    output_json[metric_name]['labels'] = [] # By default it needs the field label even with an empty list.

    for possible_label in possible_labels:
        if possible_label in parsed_json.keys():
            print("Adding {0} as label for prometheus.".format(possible_label))
            output_json[metric_name]['labels'].append(possible_label)

    
    # Some light heuristic to guess the type.
    if re.search("/time$|/bytes$|/cpu$", parsed_json['metric']):
        output_json[metric_name]["type"] = "histogram"
        output_json[metric_name]["buckets"] = [
            "10", "100", "500", "1000", "2000", "3000", "5000", "7000", "10000", "inf", "sum"
        ]
    elif re.search("/count$|/failed$|/size$", parsed_json['metric']):
        output_json[metric_name]["type"] = "gauge"
    else :
        print("Can't guess what type is this metrics: {0} Using Gauge as default type.".format(parsed_json['metric']))
        output_json[metric_name]["type"] = "gauge"
       

    # new_metric = json.dumps(output_json)
    print(json.dumps(output_json, indent=4, sort_keys=True))



